### PR TITLE
Liblinear 247

### PR DESCRIPTION
**Pull request by UnixJunkie**

### CODE REVIEW

**Comment by UnixJunkie** [comment]
At install I get a permission denied at the cp stage (there are two). But I suspect this is due to my environment, not due to the opam package. If someone else could test this package, that would be helpful.

**Comment by UnixJunkie** [comment]
related to https://github.com/ocaml/opam-repository/issues/26774

**Comment by UnixJunkie** [comment]
I'll try installing on the local disk as a test.

**Comment by UnixJunkie** [comment]
I still have a copy error if installing on the local disk

File: packages/liblinear/liblinear.2.47/opam [hunk 1]
```diff
@@ -6,8 +6,8 @@ bug-reports: "https://github.com/ocaml/opam-repository/issues"
 license: "BSD-3-Clause"
 build: [
   ["sh" "-c" "make -C liblinear-247"]
-  ["cp" "-d" "--preserve=all" "liblinear-247/train"   "%{bin}%/liblinear-train"]
-  ["cp" "-d" "--preserve=all" "liblinear-247/predict" "%{bin}%/liblinear-predict"]
+  ["install" "liblinear-247/train"   "%{bin}%/liblinear-train"]
+  ["install" "liblinear-247/predict" "%{bin}%/liblinear-predict"]
 ]
```
Comment (mseri) [inline packages/liblinear/liblinear.2.47/opam:11]: What if instead you generate a `.install` file and let opam do the copy?

**Comment by UnixJunkie** [comment]
I could try the .install file method, but everything in the opam file is
not bad either: it is centralized.

**Comment by mseri** [comment]
I was just suggesting it to see if it avoids the network disk issue

File: packages/liblinear/liblinear.2.47/opam (hunk 1 above)
Comment (mseri) [reply packages/liblinear/liblinear.2.47/opam:11]: https://opam.ocaml.org/doc/Manual.html#lt-pkgname-gt-install

**Comment by UnixJunkie** [comment]
There were some errors in the opam file (depexts instead of depends); my mistake.
But, I still get the error:
```bash
# install: cannot create regular file '~/.opam/4.14.1/bin/liblinear-train': Read-only file system
```

If someone can test on a computer that's not using NFS for the homedir, that might be helpful.

**Comment by UnixJunkie** [comment]
Is there an example of a .install file somewhere?
I did not find in opam-repos.

**Comment by shonfeder** [comment]
> Is there an example of a .install file somewhere? I did not find in opam-repos.


Perhaps https://github.com/ocaml/opam-repository/blob/56e31a3bc1fd0bfd87e5251972e806b8f78082a5/packages/xmlm/xmlm.1.1.0/opam#L31-L38 ?

I am able to find many other examples in the opam repo using `rg '\.install'` too, in case that is helpful.

File: packages/liblinear/liblinear.2.47/opam [hunk 2]
```diff
@@ -0,0 +1,35 @@
+opam-version: "2.0"
+authors: "https://www.csie.ntu.edu.tw/~cjlin/liblinear/contributors.html"
+homepage: "https://www.csie.ntu.edu.tw/~cjlin/liblinear"
+#maintainer: "https://www.csie.ntu.edu.tw/~cjlin/index.html"
+maintainer: "unixjunkie@sdf.org"
+bug-reports: "https://github.com/ocaml/opam-repository/issues"
+license: "BSD-3-Clause"
+build: [
+  ["sh" "-c" "make -C liblinear-247"]
```
Comment (mseri) [inline packages/liblinear/liblinear.2.47/opam:9]:
```suggestion
  ["sh" "-c" "make"]
```

File: packages/liblinear/liblinear.2.47/opam [hunk 3]
```diff
@@ -0,0 +1,35 @@
+opam-version: "2.0"
+authors: "https://www.csie.ntu.edu.tw/~cjlin/liblinear/contributors.html"
+homepage: "https://www.csie.ntu.edu.tw/~cjlin/liblinear"
+#maintainer: "https://www.csie.ntu.edu.tw/~cjlin/index.html"
+maintainer: "unixjunkie@sdf.org"
+bug-reports: "https://github.com/ocaml/opam-repository/issues"
+license: "BSD-3-Clause"
+build: [
+  ["sh" "-c" "make"]
```
Comment (mseri) [inline packages/liblinear/liblinear.2.47/opam:9]:
```suggestion
  [make]
```

File: packages/liblinear/liblinear.2.47/opam [hunk 4]
```diff
@@ -0,0 +1,35 @@
+opam-version: "2.0"
+authors: "https://www.csie.ntu.edu.tw/~cjlin/liblinear/contributors.html"
+homepage: "https://www.csie.ntu.edu.tw/~cjlin/liblinear"
+#maintainer: "https://www.csie.ntu.edu.tw/~cjlin/index.html"
+maintainer: "unixjunkie@sdf.org"
+bug-reports: "https://github.com/ocaml/opam-repository/issues"
+license: "BSD-3-Clause"
+build: [
+  [make]
+  ["install" "liblinear-247/train"   "%{bin}%/liblinear-train"]
```
Comment (mseri) [inline packages/liblinear/liblinear.2.47/opam:10]:
```suggestion
  ["install" "train"   "%{bin}%/liblinear-train"]
```

File: packages/liblinear/liblinear.2.47/opam [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+authors: "https://www.csie.ntu.edu.tw/~cjlin/liblinear/contributors.html"
+homepage: "https://www.csie.ntu.edu.tw/~cjlin/liblinear"
+#maintainer: "https://www.csie.ntu.edu.tw/~cjlin/index.html"
+maintainer: "unixjunkie@sdf.org"
+bug-reports: "https://github.com/ocaml/opam-repository/issues"
+license: "BSD-3-Clause"
+build: [
+  [make]
+  ["install" "train"   "%{bin}%/liblinear-train"]
+  ["install" "predict" "%{bin}%/liblinear-predict"]
```
Comment (mseri) [inline packages/liblinear/liblinear.2.47/opam:11]:
```suggestion
  ["sh" "-exc" "echo 'bin: [\\n  \"train\" {\"liblinear-train\"}\\n  \"predict\" {\"liblinear-predict\"}\\n]' > liblinear.install"]
```

Comment (mseri) [reply packages/liblinear/liblinear.2.47/opam]: It is still not working for me either. Let's try with opam's installer

**Comment by mseri** [comment]
@UnixJunkie this works!!

**Comment by mseri** [comment]
Let’s wait a moment. @UnixJunkie does this version work with NFS?

**Comment by shonfeder** [comment]
Kindly ping to, @UnixJunkie.

**Comment by UnixJunkie** [comment]
This should be merged.
I know there is a problem w/ the NFS I am forced to use, but this has nothing to do w/ this clean opam package.
The fact that opam can install liblinear in user-space is a step forward to make packages relying on liblinear easier to install (there is the linwrap package for example).

**Comment by UnixJunkie** [comment]
Once this is merged, I will replicate the recipe to create a user-space installer for libsvm.
For people who don't know, liblinear is kind of a chainsaw for machine-learning.
Libsvm is another very famous library (by the same author/research group at Taiwan national university) for the same kind of tasks, is a little bit slower but sometimes allows to get better models.

**Comment by shonfeder** [comment]
Lovely. Thanks for all your work to make this available to the community!

**Pull request merged**